Allow nested attributes

brentdodell opened the pull request:
- I don't know that I loved how this turned out...especially the last
commit. Would love some feedback @clintmiller.

brentdodell committed:
Allow filtered rspec runs (i.e. :focus)

brentdodell committed:
Prep AssociationBuilder for nested attributes

- By pulling out the way we get the primary_key & association_attributes
into methods specifying that this is the prefixed way of doing it. Next,
we'll introduce the nested way.

brentdodell committed:
Add methods for getting nested attributes

- In AssociationBuilder

brentdodell committed:
Support nested attributes in AssociationBuilder

- Nested attributes have priority over prefixed attributes
  - If we have the following, then first_name will be "Mo"
    {
      user_first_name: 'test',
      user: {
        first_name: 'Mo',
      },
    }

brentdodell committed:
Add specs to test nested attribute behavior

brentdodell committed:
Delete attributes after setting up associations

- If we don't do this, then they end up getting overwritten when super
is called in Forminate.initialize
  - I don't really like this. I would not expect AssociationBuilder to
  modify the hash that is passed in. It actually surprised me that it
  worked. I guess in Ruby, everything is passed by reference?
  - I'm looking for suggestions on a better way to do this.

clintmiller reviewed (commented):
Ruby is pass-by-value, but the values are all references. 😏 https://robertheaton.com/2014/07/22/is-ruby-pass-by-reference-or-pass-by-value/

The last commit is definitely interesting. I've got to look more into it to figure out why this is necessary.

brentdodell committed:
Move cleanup of attributes out of AssociationBldr

- Moved into Forminate#build_associations
- Ideally this would live in Forminate.initialize, since the reason
we're doing this cleanup is to prevent things from being overwritten
when super is called in .initialize...they overwritten by ActiveAttr

brentdodell committed:
Normalized keys to symbol when cleaning up

- String keys were being left behind and overwriting the attributes when
super was called.
